---
output:
  word_document: default
  html_document: default
---

```{r}
#these libraries are ported here in case you remove the code and place elsewhere.
# install.packages('Signac','Seurat') # These packages often appear to be outdated/broken, so this line ensures up to date libraries before code running. 


#Load Libraries
library(dplyr)
library(Seurat)
library(patchwork)
library(data.table)
library(stringr)
library(ggplot2)
library(ggpubr)

# set working directory
setwd("~/OneDrive - Queen Mary, University of London/QMUL/Lab/Coding/data/R/Seurat/SandLKerato")

# load .rds file with pre-filtered data. 
kerato <- readRDS(file = "smalllargekera.rds")

```
```{r}
# From https://scrnaseq-course.cog.sanger.ac.uk/website/seurat-chapter.html

# find all markers of cluster 1
#Seurat can help you find markers that define clusters via differential expression. By default, it identifes positive and negative markers of a single cluster (specified in ident.1), compared to all other cells. FindAllMarkers automates this process for all clusters, but you can also test groups of clusters vs. each other, or against all cells.

#The min.pct argument requires a feature to be detected at a minimum percentage in either of the two groups of cells, and the thresh.test argument requires a feature to be differentially expressed (on average) by some amount between the two groups. You can set both of these to 0, but with a dramatic increase in time - since this will test a large number of features that are unlikely to be highly discriminatory. As another option to speed up these computations, max.cells.per.ident can be set. This will downsample each identity class to have no more cells than whatever this is set to. While there is generally going to be a loss in power, the speed increases can be significiant and the most highly differentially expressed features will likely still rise to the top.

#percent_cutoff modulates the cutoff for the FindMarkers function
percent_cutoff = 0.25

#all cluster markers
# cluster0.markers <- FindMarkers(kerato, ident.1 = 0, min.pct = percent_cutoff)
# cluster1.markers <- FindMarkers(kerato, ident.1 = 1, min.pct = percent_cutoff)
# cluster2.markers <- FindMarkers(kerato, ident.1 = 2, min.pct = percent_cutoff)
# cluster3.markers <- FindMarkers(kerato, ident.1 = 3, min.pct = percent_cutoff)
# head(cluster1.markers, n = 5)
#cluster0.markers <- FindMarkers(kerato, ident.1 = 1, min.pct = 0.25)
#cluster2.markers <- FindMarkers(kerato, ident.1 = 1, min.pct = 0.25)
#need to check ident.1 identity - Ryan 
#write.xlsx(cluster0.markers,"cluster0_markers.xlsx")
#write.xlsx(cluster1.markers,"cluster1_markers.xlsx")
#write.xlsx(cluster2.markers,"cluster2_markers.xlsx")
#write.xlsx(cluster3.markers,"cluster3_markers.xlsx")
```

```{r}
DimPlot(kerato, reduction = "pca", pt.size = 0.5, group.by = 'none')
```


```{r}
# find markers for every cluster compared to all remaining cells, report only the positive ones
kerato.markers <- FindAllMarkers(kerato, only.pos = FALSE, min.pct = 0.25, logfc.threshold = 0.25)
kerato.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_log2FC)

cluster_markers <- kerato.markers %>% group_by(cluster) %>% filter(!between(avg_log2FC, -0.6, 0.6) & p_val_adj < 0.05)


top_markers <- kerato.markers %>% group_by(cluster) %>% filter(!between(avg_log2FC, -1, 1) & p_val_adj < 0.05)

#write out markers to excel file. Cosnider using CSV as these are easier to manipulate.
library(openxlsx)
#write.xlsx(top_markers,"allcluster_markers.xlsx")
# https://scrnaseq-course.cog.sanger.ac.uk/website/seurat-chapter.html
# has a tutorial explaining these things



```

```{r}
# We include several tools for visualizing marker expression. VlnPlot (shows expression probability distributions across clusters), and FeaturePlot (visualizes feature expression on a tSNE or PCA plot) are our most commonly used visualizations. We also suggest exploring RidgePlot, CellScatter, and DotPlot as additional methods to view your dataset. Uncomment below with genes in keratinocytes to use VlnPlot
VlnPlot(kerato, features = c("Krt10", "Krt5"))
VlnPlot(kerato, features = c("Ddx6"))

# you can plot raw counts as well
# VlnPlot(pbmc, features = c("NKG7", "PF4"), slot = "counts", log = TRUE)
```


```{r}
#build finding gene in: VariableFeatures(kerato)
# genelist <- kerato.data 
# genelist <- as.data.frame(as.matrix(genelist))
# setDT(genelist, keep.rownames = TRUE)
# genelist <- genelist[ ,1]
# 
# in_genes <- c()
#in_genes makes a list of genes and 
```

```{r}
#searching for gene in gene list
# at current this only saves the list from the R session. Must note down all genes somewhere else for now. Get it to save a file. 


# pat <- 'Sprr1a'
# 
# if (str_detect(genelist, pat) == TRUE) {
#   in_genes <- c(in_genes, pat)
# }
# 
# in_genes
```



```{r}
#we can use dot plot to show the amount of expression, and average expression of the genes within the clusters.
# https://rpubs.com/kshekhar/349874
# DotPlot(object = kerato, features = "percent.mt" , cols = c("purple", "yellow"))
```


```{r}
#reading in the correct dimensions for umap pngs. 
#setwd("~/OneDrive - Queen Mary, University of London/QMUL/Lab/Coding/data/R/Seurat/SandLKerato/images")
# library(png)
# img <- readPNG("idealsize.png")
# idealimg_dims <- dim(img) 



```


```{r}
# #attempting to automate printing our heatmap pngs. More ggplot2 knowledge will be required to sort formatting. 
# setwd("~/OneDrive - Queen Mary, University of London/QMUL/Lab/Coding/data/R/Seurat/SandLKerato/images")
# list_of_features <- c("Krt10", "Krt5", "Evpl", "Dsp")
# 
# for (gene in list_of_features){
# 
#   p <- (FeaturePlot(kerato, features = (gene)))
#   p <-  lapply(X = p, FUN = function(x) x + theme(plot.title = element_text(size = 5)))
#   png(filename = sprintf("%s_featureplot.png", gene), width = idealimg_dims[2], height = idealimg_dims[1])
#   print(p)
# 
#   dev.off()
# }
```


```{r}
#set working directory
setwd("~/OneDrive - Queen Mary, University of London/QMUL/Lab/Coding/data/R/Seurat/SandLKerato")

#assign cluster IDs - two separate statements depending on what you want, numerics or character labels
new.cluster.ids <- c("Metabolism and Respiration", "Development and Reduced Translation", "Translating", "Keratinocyte Differentiation")

#new.cluster.ids <- c(0,1,2,3)
names(new.cluster.ids) <- levels(kerato)
kerato <- RenameIdents(kerato, new.cluster.ids)
DimPlot(kerato, reduction = "umap", label = TRUE, pt.size = 0.5) + NoLegend()

# number and proportion of cells in each cluster
# from https://satijalab.org/seurat/v3.0/interaction_vignette.html
table(Idents(kerato))
prop.table(table(Idents(kerato)))
```
```{r}
#if we want to subset this analysis based off of this data, we can do so with this tutorial: https://satijalab.org/seurat/v3.0/interaction_vignette.html

# WhichCells(pbmc, idents = "NK")

# How can I extract expression matrix for all NK cells (perhaps, to load into another package)
# nk.raw.data <- as.matrix(GetAssayData(pbmc, slot = "counts")[, WhichCells(pbmc, ident = "NK")])
# 
# # Can I create a Seurat object based on expression of a feature or value in object metadata?
# subset(pbmc, subset = MS4A1 > 1)

# Can I create a Seurat object of all cells except the NK cells and B cells?
# subset(pbmc, idents = c("NK", "B"), invert = TRUE)
```


```{r}
# Monocle Analysis 
# sajita lab have a tutorial on using monocle from seurat objects
# https://satijalab.org/signac/articles/monocle.html

setwd("~/OneDrive - Queen Mary, University of London/QMUL/Lab/Coding/data/R/Seurat/SandLKerato")

#loading libraries
library(Signac)
library(Seurat)
library(SeuratWrappers)
library(monocle3)
library(Matrix)
library(ggplot2)
library(patchwork)
library(tidyseurat)
set.seed(1234)
```




```{r}
# this piece of code may let us subset the data, lifted from tutorial
# erythroid <- bone[,  bone$assigned_celltype %in% c("HSC", "MEP", "CMP-BMP")]
# kerato cluster IDs: c("Metabolism and Respiration", "Development and Reduced Translation", "Translating", "Keratinocyte Differentiation")

# this will select the subset of the clusters using their ID number
# kerato[,  kerato@meta.data[["seurat_clusters"]] %in% c("1", "2","3")]
# just make a new variable and run it through the analysis. Alternatively, subset the seurat object, recluster and go from there. 
kerato <- readRDS(file = "smalllargekera.rds")
kerato.cds <- as.cell_data_set(kerato)
kerato.cds <- cluster_cells(cds = kerato.cds, reduction_method = "UMAP")
kerato.cds <- learn_graph(kerato.cds, use_partition = TRUE)
```

```{r}
# order cells
root_group = colnames(kerato.cds)[clusters(kerato.cds) == 1]

kerato.cds <- order_cells(kerato.cds, reduction_method = "UMAP", root_cells = root_group)

# plot trajectories colored by pseudotime
plot_cells(
  cds = kerato.cds,
  color_cells_by = "pseudotime",
  show_trajectory_graph = TRUE
)

# add monocle pseudotime data to seurat metadata

kerato <- AddMetaData(
  object = kerato,
  metadata = kerato.cds@principal_graph_aux@listData$UMAP$pseudotime,
  col.name = "Pseudotime"
)

saveRDS(kerato, file = "pseudoSLkera.rds")

FeaturePlot(kerato, features = "Pseudotime", label = TRUE)
#FeaturePlot(kerato, features = "Pseudotime", cols = c("purple", "yellow")) 
#FeaturePlot(kerato, features = "Pseudotime") & scale_color_viridis_c()


```

```{r}
#internal monocle function will print citations used. 
get_citations(kerato.cds)
```









